middle_end/flambda2/types/grammar/type_grammar.ml

```diff
@@ -521,18 +523,19 @@ and apply_renaming_head_of_kind_value head renaming =
         alloc_mode
       } ->
     let length' = apply_renaming length renaming in
-    let fields' =
+    let fields_list = Array.to_list fields in
+    let fields_list' =
       Misc.Stdlib.List.map_sharing
         (fun field -> apply_renaming field renaming)
-        fields
+        fields_list
     in
-    if length == length' && fields == fields'
+    if length == length' && fields_list == fields_list'
```

Comment:
I think you could just use `Array.map` and then `Array.forall2 (==)` to detect equality.